### PR TITLE
[Accessibility] Fixing a default accessible name of a list of a dropdown entry value in PropertyGridView

### LINKED ISSUE #2802: The UIA Name property is missing on the list element in selected PropertyGrid item
* .NET Core Version: 3.1.200-preview-014883

* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**
The Name property of the focusable element(list) is null.
Actual

**Expected behavior:**
The list element should have a meaningful name instead of null.

**Minimal repro:**
1. Create a WinForm 4.8 app with a PropertyGrid whose property items contain a list of strings or run the attached core app. [UIA_Name_missing.zip](https://github.com/dotnet/winforms/files/4152306/UIA_Name_missing.zip)
2. Run the Accessibility Insights for Windows tool to examine the UIA representation of the UI.
3. When keyboard focus is over in the right side of a property item, check the UIA tree.

### PR DESCRIPTION
**Pull request by vladimir-krestov**
Fixes #2802

## Proposed changes

- Use a PropertyGrid entry name if a list AccessibleName isn't set

## Customer Impact

- A user can see a non-empty accessible name of a values list using Inspect 
(Does not affect Narrator announcing)
- Fixed MAS requirement:
image

## Regression? 

- No

## Risk

- Minimal

### Before
- An empty AccessibleName
image

### After
- A non-empty AccessibleName (<del>"{Property name}"</del> -> "{Property name} property values")
image

## Test methodology 

- CTI
- Unit testing

## Accessibility testing  
- Using Narrator, Accessibility Insights and Inspect

 

## Test environment(s) 

- .NET Core SDK: 5.0.0-alpha.1.20072.3
- Microsoft Windows [Version 10.0.18363.592]

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2843)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs [hunk 1]
```diff
@@ -7232,6 +7232,12 @@ internal override UiaCore.IRawElementProviderFragment FragmentNavigate(UiaCore.N
                 return base.FragmentNavigate(direction);
             }
 
+            public override string Name
+            {
+                get => string.IsNullOrEmpty(base.Name) ? _owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName : base.Name;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: I think that an empty string as an accessible name is the way to remove accessible name from the control.

Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]:
```suggestion
                get => base.Name ?? _owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName;
```

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Had you discussed with Merrie if property name is appropriate in this scenarioo? I would suggest "List of {PropertyName} properties." I think it's confusing as is.

Also, is it possible for the applicaiton developer to set Accessible name by some other means?

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs [hunk 2]
```diff
@@ -7232,6 +7232,12 @@ internal override UiaCore.IRawElementProviderFragment FragmentNavigate(UiaCore.N
                 return base.FragmentNavigate(direction);
             }
 
+            public override string Name
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: After the fix is finalized, consider pulling this class into a separate file in another PR.

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Okay, I'll do. Thanks

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs (hunk 1 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: @Tanya-Solyanik 
1) Good point. I'll change the implementation as you suggested below.
2) No, I hadn't discussed. I'll ask her after we find a solution.
    >I think it's confusing as is.

    I agree with you but we can't use "List of..." accessible name due to MAS requirements:
image
So I suggest using "{PropertyName} values" name and "list" control type.
image

3) >Also, is it possible for the applicaiton developer to set Accessible name by some other means?

    No, an application developer can't it. But it may be necessary for an internal developer.

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: But maybe better to use "{Property name} property values"

image

Anyway, I couldn't find a way to announce the list AccessibleName using Narrator. Though it may be fixed in the future.

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: @Tanya-Solyanik could you please confirm this solution?

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs [hunk 3]
```diff
@@ -7232,6 +7232,12 @@ internal override UiaCore.IRawElementProviderFragment FragmentNavigate(UiaCore.N
...
+            {
+                get => base.Name ?? string.Format(SR.PropertyGridEntryValuesListDefaultAccessibleName, _owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Please double check how `string.Format` handles null parameters. What would happen if `_owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName` resolves to null? 
Also this line is too long, consider re-formatting it, or not using expression-bodied property at all.

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 4]
```diff
@@ -6611,4 +6611,7 @@ Stack trace where the illegal operation occurred was:
   <data name="ErrorProviderDefaultAccessibleName" xml:space="preserve">
     <value>Error provider</value>
   </data>
+  <data name="PropertyGridEntryValuesListDefaultAccessibleName" xml:space="preserve">
+    <value>{0} property values</value>
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/Resources/SR.resx]: This accessible name looks good to me.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 5]
```diff
@@ -43,5 +44,27 @@ public void PropertyGridViewAccessibleObject_Supports_TablePattern(int pattern)
...
+            PropertyGrid propertyGrid = new PropertyGrid();
+            propertyGrid.SelectedObject = new Button();
+            Control.ControlAccessibleObject gridViewAccessibleObject = propertyGrid.GridViewAccessibleObject as Control.ControlAccessibleObject;
+            PropertyGridView propertyGridView = gridViewAccessibleObject.Owner as PropertyGridView;
+            CategoryGridEntry category = propertyGrid.GetPropEntries()[1] as CategoryGridEntry; // "Appearance"
+            GridEntry entry = category.Children[0] as GridEntry; // "BackColor"
+            propertyGrid.SelectedGridItem = entry;
+
+            string listAccessibleName = propertyGridView.DropDownListBoxAccessibleObject.Name;
+            Assert.False(string.IsNullOrEmpty(listAccessibleName));
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: this `Assert.False` seems redundant

File: src/System.Windows.Forms/src/Resources/SR.resx (hunk 4 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/Resources/SR.resx]: Good, thanks

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 6]
```diff
@@ -43,5 +44,27 @@ public void PropertyGridViewAccessibleObject_Supports_TablePattern(int pattern)
             var accessibleObject = parentGridEntry.AccessibilityObject;
             Assert.True(accessibleObject.IsPatternSupported((UiaCore.UIA)pattern));
         }
+
+        [WinFormsFact]
+        public void PropertyGridViewAccessibleObject_EntryValuesListName_IsNotEmpty()
+        {
+            PropertyGrid propertyGrid = new PropertyGrid();
+            propertyGrid.SelectedObject = new Button();
+            Control.ControlAccessibleObject gridViewAccessibleObject = propertyGrid.GridViewAccessibleObject as Control.ControlAccessibleObject;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I would use a cast operator here, explicit cast exception is preferable to a null ref on the next line.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 7]
```diff
@@ -43,5 +44,27 @@ public void PropertyGridViewAccessibleObject_Supports_TablePattern(int pattern)
...
+            PropertyGridView propertyGridView = gridViewAccessibleObject.Owner as PropertyGridView;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I would use cast here as well.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: and on lines 55, 56

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 8]
```diff
@@ -43,5 +44,27 @@ public void PropertyGridViewAccessibleObject_Supports_TablePattern(int pattern)
...
+            propertyGrid.SelectedGridItem = entry;
+
+            string listAccessibleName = propertyGridView.DropDownListBoxAccessibleObject.Name;
+            Assert.False(string.IsNullOrEmpty(listAccessibleName));
+            Assert.Equal(string.Format(SR.PropertyGridEntryValuesListDefaultAccessibleName, entry.PropertyName), listAccessibleName);
+
+            string newValue = "Some text";
+            propertyGridView.DropDownListBoxAccessibleObject.Name = newValue;
+            listAccessibleName = propertyGridView.DropDownListBoxAccessibleObject.Name;
+            Assert.False(string.IsNullOrEmpty(listAccessibleName));
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: `Assert.False` is redundant

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: the new name looks reasonable to me. Please see other comments

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs (hunk 3 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Good point! Null value throws an exception. I'll change the implementation.

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs (hunk 1 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Thanks

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 5 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I disagree. The test was created to check that MAS requirement is met:
image

Maybe in the future, someone accidentally will break this accessible name. Then the test won't be passed and it will warn about breakage.

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: @Tanya-Solyanik what do you think?

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 8 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: https://github.com/dotnet/winforms/pull/2843#discussion_r378695355

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 6 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I agree, fixed.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 7 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: Thanks, fixed.

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs (hunk 3 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Fixed

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 5 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I think that if accessible name is set to null, the test will fail on the next line when matching against the expected name.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 8 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: the test will fail on the next line.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 5 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: Yes indeed. If someone changes the implementation of getting an accessible name, he will have to change this test too.
Thank you!

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 8 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: I'll remove this. Thanks!

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: Done

**Comment by vladimir-krestov** [comment]
Will rebase and squash commits after testers approving.

**Comment by vladimir-krestov** [comment]
Testers approved the fix. ✔️ Need a final review.

**Comment by vladimir-krestov** [comment]
The PR doesn't relate to #2870 because an app developer can't change GridViewListBox AccessibleName. It is possible internally.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 9]
```diff
@@ -263,5 +263,25 @@ public void PropertyGridViewAccessibleObject_Parent_IsNotNull()
             ControlAccessibleObject accessibleObject = (ControlAccessibleObject)propertyGrid.GridViewAccessibleObject;
             Assert.NotNull(accessibleObject.Parent);
         }
+
+        [WinFormsFact]
+        public void PropertyGridViewAccessibleObject_EntryValuesListName_IsNotEmpty()
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: Which codepath is this test covering?

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: image
We have to ensure compliance with MAS responsibilities because an app developer doesn't have access to the Entry Values List AccessibleObject.
This test checks that this is done.

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: We have added a new property `Name`, but the test (according to the name) exercises `EntryValuesListName`.
So my question stems from this - are we testing the new property? 

Also the property has several codepaths, e.g. 
* 1 here: `if (base.Name != null)`
* 3 here: `_owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName`
* 1 here: `propertyName ??= string.Empty`

Which one are we testing here? And do we have tests for the other paths?

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: @RussKie 

> are we testing the new property?

Yes, we are testing propertyGridView.DropDownListBoxAccessibleObject.Name, which I implemented in this PR.

> 1 here: if (base.Name != null)

 `(base.Name != null)` means a custom accessible name was set.
It is checked here:
```cs
            string newValue = "Some text";
            propertyGridView.DropDownListBoxAccessibleObject.Name = newValue;
            listAccessibleName = propertyGridView.DropDownListBoxAccessibleObject.Name;
            Assert.Equal(newValue, listAccessibleName);
```

> - 4 here: _owningGridViewListBox?.OwningPropertyGridView?.SelectedGridEntry?.PropertyName
> - 1 here: propertyName ??= string.Empty

In any negative case, the string will be empty or null.
We could use `Assert.False(string.IsNullOrEmpty(listAccessibleName));` but I removed it as we decided in https://github.com/dotnet/winforms/pull/2843#discussion_r378684827.

> Which one are we testing here? And do we have tests for the other paths?

This test tests all code paths at the same time.

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: > Yes, we are testing propertyGridView.DropDownListBoxAccessibleObject.Name, which I implemented in this PR.

Please rename the test to reflect what's being tested in the name, something like:
```suggestion
        public void PropertyGridView_DropDownListBoxAccessibleObject_Name_EntryValuesListName_IsExpected()
```

> This test tests all code paths at the same time.

Each unit test must test only a single code path.

> In any negative case, the string will be empty or null.

Correct, but each `?.` is a shortcut for `!= null`. If tomorrow one of these `?` gets accidentally deleted how do catch it?

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs [hunk 10]
```diff
@@ -7232,6 +7232,26 @@ internal override UiaCore.IRawElementProviderFragment FragmentNavigate(UiaCore.N
                 return base.FragmentNavigate(direction);
             }
 
+            public override string Name
+            {
+                get
+                {
+                    if (base.Name != null)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: Consider introducing a local variable to hold `base.Name`

Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs]: @Tanya-Solyanik, I have changed the implementation. Please check and approve.

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 11]
```diff
@@ -264,24 +264,35 @@ public void PropertyGridViewAccessibleObject_Parent_IsNotNull()
             Assert.NotNull(accessibleObject.Parent);
         }
 
+        [Theory]
+        [InlineData("Some test text")]
+        [InlineData("")]
+        public void PropertyGridView_GridViewListBoxAccessibleObject_Name_ReturnsDeterminedName(string name)
+        {
+            using PropertyGrid propertyGrid = new PropertyGrid();
+            propertyGrid.SelectedObject = new Button();
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]:
```suggestion
            using PropertyGrid propertyGrid = new PropertyGrid();
            using Button button = new Button();
            propertyGrid.SelectedObject = button;
```

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs [hunk 12]
```diff
@@ -264,24 +264,35 @@ public void PropertyGridViewAccessibleObject_Parent_IsNotNull()
...
+            Assert.Equal(name, listAccessibleName);
+        }
+
         [WinFormsFact]
-        public void PropertyGridViewAccessibleObject_EntryValuesListName_IsNotEmpty()
+        public void PropertyGridView_GridViewListBoxAccessibleObject_ReturnsDefaultName_IfBaseNameIsNull()
         {
-            PropertyGrid propertyGrid = new PropertyGrid();
+            using PropertyGrid propertyGrid = new PropertyGrid();
             propertyGrid.SelectedObject = new Button();
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: ditto

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 9 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: @RussKie, due the implementation changing, I haven't to make a lot of tests to check it. It needs 2 tests only now.
Thanks for your recommendations!

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: @RussKie, due to the implementation changing, I haven't to make a lot of tests to check it. It needs 3 tests only now.
Thanks for your recommendations!

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 11 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: It's not needed anymore. Thanks

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs (hunk 12 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewAccessibleObjectTests.cs]: Fixed in fdb9cdede03a8f3c689f99fe8376c55628324005 commit. Thanks

**Comment by vladimir-krestov** [comment]
Testers approved the fix. Automation tests are passed, no regression issue found. ✔️
